picoquic/frames.c

```diff
@@ -220,56 +268,89 @@ int picoquic_prepare_stream_reset_frame(picoquic_stream_head* stream,
     return ret;
 }
 
-int picoquic_decode_stream_reset_frame(picoquic_cnx_t* cnx, uint8_t* bytes,
-    size_t bytes_max, size_t* consumed)
+
+static int picoquic_decode_padding_or_ping_frame(picoquic_cnx_t* cnx, const uint8_t** pbytes, const uint8_t* bytes_max,
+                                                    picoquic_frame_type_enum_t frame_type)
+{
+    /* As an optimization, consume multiple subsequent padding in one go */
+    while (*pbytes < bytes_max && **pbytes == frame_type) {
+        ++*pbytes;
+    }
+    return 0;
+}
+
+
+int picoquic_decode_padding_frame(picoquic_cnx_t* cnx, const uint8_t** pbytes, const uint8_t* bytes_max, int restricted, uint64_t current_time)
+{
+    return picoquic_decode_padding_or_ping_frame(cnx, pbytes, bytes_max, picoquic_frame_type_padding);
+}
+
+int picoquic_decode_ping_frame(picoquic_cnx_t* cnx, const uint8_t** pbytes, const uint8_t* bytes_max, int restricted, uint64_t current_time)
+{
+    return picoquic_decode_padding_or_ping_frame(cnx, pbytes, bytes_max, picoquic_frame_type_ping);
+}
+
+int picoquic_decode_blocked_frame(picoquic_cnx_t* cnx, const uint8_t** pbytes, const uint8_t* bytes_max, int restricted, uint64_t current_time)
+{
+    /* Skip the max data offset */
+    return picoquic_varint_skip_check(cnx, pbytes, bytes_max);
+}
+
+int picoquic_decode_stream_blocked_frame(picoquic_cnx_t* cnx, const uint8_t** pbytes, const uint8_t* bytes_max, int restricted, uint64_t current_time)
+{
+    /* TODO: check that the stream number is valid */
+    int ret = picoquic_varint_skip_check(cnx, pbytes, bytes_max);
+    if (ret == 0) {
+        /* Skip the max data offset */
+        ret = picoquic_varint_skip_check(cnx, pbytes, bytes_max);
+    }
+    return ret;
+}
+
+int picoquic_decode_stream_id_needed_frame(picoquic_cnx_t* cnx, const uint8_t** pbytes, const uint8_t* bytes_max, int restricted, uint64_t current_time)
+{
+    return picoquic_varint_skip_check(cnx, pbytes, bytes_max);
+}
+
```

Comment:
I was trying to optimize human-readable code, not machine code.  The human-readable code does get smaller (while the extra indirection will likely be optimized away by compiler and HW).

It is possible to switch to
   next_pointer = (cnx, current_pointer, max_pointer, &ret, ...)
It is a bit unusual (usually return is a status is the return of the function), but it can work.  It does add an extra argument to the function.

> Also, I am dead set against the multiple return structure. It is a form of goto, and I insist on using structured code instead.

Your position regarding returns (and, implicitly, goto) is clear.  What about "breaks" and "continues"?

(My preference is to hold having the code follow the "natural" logic of the process encoded as much as possible.  I.e. if someone asked me to do some multi-step task, and I find early on that I cannot finish, I would quit right away and tell them what happened right there and then.  I'd rather bend the coding style to align with thought flow than the other way around.  But, yes, one coding style can be converted into the other.)